Document staging scripts (reference) (#356)
epinio/docs: docs/references/customization/staging-scripts.md

---

> title: ""
> sidebar_label: "Staging Scripts"

Reviewer: ```suggestion
title: "Customization point: Staging scripts"
sidebar_label: "Staging scripts"
description: Using staging scripts with Epinio.
keywords: [epinio, application development, staging scripts, kubernetes]
```

---

# Customization point: Staging Scripts

## Overview

:::note
This customization interacts with the
[customization of buildpacks](staging.md).
:::

> Staging scripts are what Epinio uses to interface with the
> Paketo [Cloud Native Buildpacks](https://buildpacks.io/).

Reviewer: ```suggestion
Epinio uses staging scripts to interact with Paketo [Cloud Native Buildpacks](https://buildpacks.io/).
```

Epinio automatically selects the set of staging scripts based on the name of the chosen builder
image, and the images supported by a specific definition.

By default Epinio installs three definitions.

>   1. One to support Bionic-based builder images (`paketo-buildpacks/builder:*`).
>   2. One to support Jammy-based  builder images (`paketo-buildpacks/builder-jammy-*:*`).
>   3. And one to serve as fallback for any images not captured by the other two.

Reviewer: ```suggestion
  1. One to support Bionic-based builder images (`paketo-buildpacks/builder:*`).
  1. One to support Jammy-based  builder images (`paketo-buildpacks/builder-jammy-*:*`).
  1. One to serve as fallback for any images not captured by the other two.
```

Reviewer 2: Is it not better to use bullet points instead of a numeric list in this case?

Author: It might be a style-preference. I liked numeric list better for this.

|---	  	|---							|
|`PROTOCOL`	|S3 storage, protocol					|
|`ENDPOINT`	|S3 storage, primary endpoint				|
|`BUCKET`	|S3 storage, name of bucket to read			|
|`BLOBID`	|S3 storage, blod id/name of app sources, in the bucket	|

### Staging script API, Unpack

The `unpack` script is executed using the `unpackImage` and further configured
with a single environment variable:

|Name	  	|Content							|
|---	  	|---								|
|`BLOBID`	|blob id / file name of the app source archive to unpack	|
|`USERID`	|Numeric id of the `cnb` user used to run the `build` script	|
|`GROUPID`	|Numeric id of the user group used to run the `build` script	|

### Staging script API, Build

The `build` script is executed using the using the chosen builder image and further configured with
a two environment variables:

|Name	  	|Content					|
|---	  	|---						|
|`PREIMAGE`	|url to the result image from a previous push	|
|`APPIMAGE`	|url to save the new application image under	|
|`USERID`	|Numeric id of the `cnb` user used to run the `build` script	|
|`GROUPID`	|Numeric id of the user group used to run the `build` script	|

When present the `PREIMAGE` is used by `/cnb/lifecycle/creator` as a cache for layers,

> speeding compilation up.

Reviewer: ```suggestion
speeding up compilation.
```

## Overview

:::note
This customization interacts with the
[customization of buildpacks](staging.md).
:::

Epinio uses staging scripts to interact with Paketo [Cloud Native Buildpacks](https://buildpacks.io/).

Epinio automatically selects the set of staging scripts based on the name of the chosen builder
image, and the images supported by a specific definition.

By default Epinio installs three definitions.

  1. One to support Bionic-based builder images (`paketo-buildpacks/builder:*`).
  1. One to support Jammy-based  builder images (`paketo-buildpacks/builder-jammy-*:*`).
  1. One to serve as fallback for any images not captured by the other two.

It is this last definition which is configured when [customizing buildpacks](staging.md)
with a different image.

By default this fallback is configured for Jammy-based images in general, and the
[jammy full stack paketo builder image](https://github.com/paketo-buildpacks/builder-jammy-full)
in particular

## Specification

A set of staging scripts is represented by a kubernetes ConfigMap resource labeled
with `app.kubernetes.io/component: epinio-staging`.

This ConfigMap is expected to have the following keys:

> |Key		|Content   |
> |---		|---	   |
> |`builder`	|Glob-pattern to match the image references supported by this spec.	|
> |`userID`	|User ID to run the `build` script with.				|
> |`groupID`	|Group ID to run the `build` script with.				|
> |||
> |`base`		|Optional redirect to the actual ConfigMap with the scripts.		|
> |||
> |`build`	|Shell script to compile the unpacked sources into an image.		|
> |		|Executed using the chosen builder image      	      			|
> |`downloadImage`|Container image to run the `download` script with.			|
> |`download`	|Shell script to retrieve the app sources from the Epinio's S3 storage.	|
> |		|Executed using the `downloadImage`.      	      			|
> |`unpackImage`	|Container image to run the `unpack` script with.			|
> |`unpack`	|Shell script to unpack the app sources into a directory tree.		|
> |		|Executed using the `unpackImage`.      	      			|

Reviewer 2: ```suggestion
|Key		|Content   |
|---		|---	   |
|`builder`	|Glob-pattern to match the image references supported by this spec.	|
|`userID`	|User ID to run the `build` script with.				|
|`groupID`	|Group ID to run the `build` script with.				|
|`base`		|Optional redirect to the actual ConfigMap with the scripts.		|
|`build`	|Shell script to compile the unpacked sources into an image.<br/>Executed using the chosen builder image |
|`downloadImage`|Container image to run the `download` script with.			|
|`download`	|Shell script to retrieve the app sources from the Epinio's S3 storage.<br/>Executed using the `downloadImage`.	|
|`unpackImage`	|Container image to run the `unpack` script with.			|
|`unpack`	|Shell script to unpack the app sources into a directory tree.<br/>Executed using the `unpackImage`. |
```

### result

|Key		|Content   |
|---		|---	   |
|`builder`	|Glob-pattern to match the image references supported by this spec.	|
|`userID`	|User ID to run the `build` script with.				|
|`groupID`	|Group ID to run the `build` script with.				|
|`base`		|Optional redirect to the actual ConfigMap with the scripts.		|
|`build`	|Shell script to compile the unpacked sources into an image.<br/>Executed using the chosen builder image |
|`downloadImage`|Container image to run the `download` script with.			|
|`download`	|Shell script to retrieve the app sources from the Epinio's S3 storage.<br/>Executed using the `downloadImage`.	|
|`unpackImage`	|Container image to run the `unpack` script with.			|
|`unpack`	|Shell script to unpack the app sources into a directory tree.<br/>Executed using the `unpackImage`. |

Author: I intentionally used empty table rows to visually separate the logical groups of keys in a single table, i.e
- match related, and spec specific
- indirection, base spec
- script information the base spec can override.

I did not like to have three tables. And having them together without separation feels wrong to me.
Maybe place rows with some kind of subtitling after the empty ones?

Reviewer 2: Ah, actually what I didn't like was the script description in a new line (with the empty key).

![image](https://github.com/epinio/docs/assets/1763949/549129b5-f82f-4ee8-945a-a0795fab4bc3)

I'm "ok" keeping the empty rows for grouping, but I still prefer to use the `<br/>` for the scripts info, if you are ok with it:

|Key		|Content   |
|---		|---	   |
|`builder`	|Glob-pattern to match the image references supported by this spec.	|
|`userID`	|User ID to run the `build` script with.				|
|`groupID`	|Group ID to run the `build` script with.				|
|||
|`base`		|Optional redirect to the actual ConfigMap with the scripts.		|
|||
|`build`	|Shell script to compile the unpacked sources into an image.<br/>Executed using the chosen builder image |
|`downloadImage`|Container image to run the `download` script with.			|
|`download`	|Shell script to retrieve the app sources from the Epinio's S3 storage.<br/>Executed using the `downloadImage`.	|
|`unpackImage`	|Container image to run the `unpack` script with.			|
|`unpack`	|Shell script to unpack the app sources into a directory tree.<br/>Executed using the `unpackImage`. |

Author: Ah. Yes, that is ok. Will do that now.

Author: Done with commit 48f354c1a9eab236bb326d966b3b527832992c0c

Author: Oh. This needs more work. The CI reminds us that docasaurus is actually not markdown in the core.

Author: Ah, wheww. Only a typo in one of the new tags.
